File: apps/ios/Tests/CrossyProtocolTests/ReactionShapeTests.swift
import Foundation
import XCTest

import CrossyProtocol

// Emoji reactions: shape only, receive-any (PROTOCOL.md §5, §6, §9). Twinned
// case-for-case from the reactions describe block in packages/protocol/src/codec.test.ts:
// the decoders enforce a non-empty string of at most 32 UTF-8 bytes and NEVER set
// membership, which is session-service policy, so an emoji outside the v1 set decodes
// in both directions and the published set can widen without a version bump (§14).

final class ReactionShapeTests: XCTestCase {
    // MARK: - Receive-any, send-gated (§9)

    func test_decodesAReactWhoseEmojiIsOutsideTheV1Set_receiveAny_PROTOCOL9() throws {
        // The codec checks shape, not set membership; gating is the sender's policy.
        let frame = Data(#"{"type":"react","emoji":"🔥","cell":3}"#.utf8)
        let decoded = try JSONDecoder().decode(ClientMessage.self, from: frame)
        XCTAssertEqual(decoded, .react(ReactMessage(emoji: "🔥", cell: 3)))
    }

    func test_decodesAReactionWhoseEmojiIsOutsideTheV1Set_receiveAny_PROTOCOL9() throws {
        // A receiver MUST NOT reject an unknown emoji (receive-any, send-gated, §9).
        let frame = Data(#"{"type":"reaction","userId":"u2","emoji":"🦀","cell":3}"#.utf8)
        let decoded = try JSONDecoder().decode(ServerMessage.self, from: frame)
        XCTAssertEqual(decoded, .reaction(ReactionMessage(userId: "u2", emoji: "🦀", cell: 3)))
    }

    // MARK: - Forward compatibility (§3)

    func test_ignoresUnknownExtraFieldsOnReactAndReaction_PROTOCOL3() throws {
        let react = Data(
            #"{"type":"react","emoji":"🎉","cell":3,"futureField":{"nested":true}}"#.utf8)
        let decodedReact = try JSONDecoder().decode(ClientMessage.self, from: react)
        XCTAssertEqual(decodedReact, .react(ReactMessage(emoji: "🎉", cell: 3)))
        let reencodedReact = try XCTUnwrap(
            try jsonObject(JSONEncoder().encode(decodedReact)) as? NSDictionary)
        XCTAssertNil(reencodedReact["futureField"])
        XCTAssertEqual(
            Set(reencodedReact.allKeys.compactMap { $0 as? String }),
            ["type", "emoji", "cell"])

        let reaction = Data(
            #"{"type":"reaction","userId":"u2","emoji":"🎉","cell":3,"futureField":1}"#.utf8)
        let decodedReaction = try JSONDecoder().decode(ServerMessage.self, from: reaction)
        XCTAssertEqual(
            decodedReaction, .reaction(ReactionMessage(userId: "u2", emoji: "🎉", cell: 3)))
        let reencodedReaction = try XCTUnwrap(
            try jsonObject(JSONEncoder().encode(decodedReaction)) as? NSDictionary)
        XCTAssertNil(reencodedReaction["futureField"])
    }

    // MARK: - Malformed frames (§5, §11 drop-and-log posture)

    func test_rejectsAReactMissingEmojiAsMalformed_PROTOCOL5() {
        let frame = Data(#"{"type":"react","cell":3}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ClientMessage.self, from: frame)) { error in
            XCTAssertTrue(error is DecodingError, "missing emoji is malformed, got \(error)")
        }
    }

    func test_rejectsAReactionWithAMistypedCellAsMalformed_PROTOCOL6() {
        let frame = Data(#"{"type":"reaction","userId":"u2","emoji":"🎉","cell":"3"}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ServerMessage.self, from: frame)) { error in
            XCTAssertTrue(error is DecodingError, "a string cell is malformed, got \(error)")
        }
    }

    // MARK: - The 32-byte shape rule (§9)

    func test_rejectsAnEmptyEmojiAsMalformed_PROTOCOL9() {
        let frame = Data(#"{"type":"react","emoji":"","cell":3}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ClientMessage.self, from: frame)) { error in
            XCTAssertTrue(error is DecodingError, "empty emoji is malformed, got \(error)")
        }
    }

    func test_rejectsAnEmojiOver32UTF8BytesAsMalformed_PROTOCOL9() {
        // Nine 🎉 graphemes are 36 UTF-8 bytes (4 each), past the 32-byte shape cap.
        let frame = Data(
            #"{"type":"react","emoji":"\#(String(repeating: "🎉", count: 9))","cell":3}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ClientMessage.self, from: frame)) { error in
            XCTAssertTrue(error is DecodingError, "36 bytes is malformed, got \(error)")
        }
    }

    func test_acceptsAnEmojiExactlyAtThe32UTF8ByteCap_PROTOCOL9() throws {
        // Eight 🎉 graphemes are exactly 32 UTF-8 bytes, the inclusive boundary.
        let emoji = String(repeating: "🎉", count: 8)
        let frame = Data(#"{"type":"react","emoji":"\#(emoji)","cell":3}"#.utf8)
        let decoded = try JSONDecoder().decode(ClientMessage.self, from: frame)
        XCTAssertEqual(decoded, .react(ReactMessage(emoji: emoji, cell: 3)))
    }

    func test_reactionDecoderAppliesTheSameShapeRule_PROTOCOL9() {
        // The inbound notice enforces shape exactly as the outbound command does:
        // one rule, both directions (the codec's single asEmoji).
        let empty = Data(#"{"type":"reaction","userId":"u2","emoji":"","cell":3}"#.utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ServerMessage.self, from: empty))
        let oversize = Data(
            #"{"type":"reaction","userId":"u2","emoji":"\#(String(repeating: "🎉", count: 9))","cell":3}"#
                .utf8)
        XCTAssertThrowsError(try JSONDecoder().decode(ServerMessage.self, from: oversize))
    }
}

File: apps/android/protocol/src/test/kotlin/crossy/protocol/ReactionShapeTests.kt
package crossy.protocol

import kotlinx.serialization.SerializationException
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

// Emoji reactions: shape only, receive-any (PROTOCOL.md §5, §6, §9). Twinned case-for-case from
// apps/ios ReactionShapeTests.swift and the reactions describe block in
// packages/protocol/src/codec.test.ts: the decoders enforce a non-empty string of at most 32 UTF-8
// bytes and NEVER set membership, which is session-service policy, so an emoji outside the v1 set
// decodes in both directions and the published set can widen without a version bump (§14).
class ReactionShapeTests {
    // --- Receive-any, send-gated (§9) ---

    @Test
    fun decodesAReactWhoseEmojiIsOutsideTheV1Set_receiveAny_PROTOCOL9() {
        // The codec checks shape, not set membership; gating is the sender's policy.
        val frame = """{"type":"react","emoji":"🔥","cell":3}"""
        val decoded = ProtocolJson.decodeFromString(ClientMessageSerializer, frame)
        assertEquals(ClientMessage.React(ReactMessage(emoji = "🔥", cell = 3)), decoded)
    }

    @Test
    fun decodesAReactionWhoseEmojiIsOutsideTheV1Set_receiveAny_PROTOCOL9() {
        // A receiver MUST NOT reject an unknown emoji (receive-any, send-gated, §9).
        val frame = """{"type":"reaction","userId":"u2","emoji":"🦀","cell":3}"""
        val decoded = ProtocolJson.decodeFromString(ServerMessageSerializer, frame)
        assertEquals(ServerMessage.Reaction(ReactionMessage(userId = "u2", emoji = "🦀", cell = 3)), decoded)
    }

    // --- Forward compatibility (§3) ---

    @Test
    fun ignoresUnknownExtraFieldsOnReactAndReaction_PROTOCOL3() {
        val react = """{"type":"react","emoji":"🎉","cell":3,"futureField":{"nested":true}}"""
        val decodedReact = ProtocolJson.decodeFromString(ClientMessageSerializer, react)
        assertEquals(ClientMessage.React(ReactMessage(emoji = "🎉", cell = 3)), decodedReact)
        val reencodedReact = ProtocolJson.parseToJsonElement(
            ProtocolJson.encodeToString(ClientMessageSerializer, decodedReact),
        )
        assertEquals(setOf("type", "emoji", "cell"), allJsonKeys(reencodedReact).toSet())

        val reaction = """{"type":"reaction","userId":"u2","emoji":"🎉","cell":3,"futureField":1}"""
        val decodedReaction = ProtocolJson.decodeFromString(ServerMessageSerializer, reaction)
        assertEquals(
            ServerMessage.Reaction(ReactionMessage(userId = "u2", emoji = "🎉", cell = 3)),
            decodedReaction,
        )
        val reencodedReaction = ProtocolJson.parseToJsonElement(
            ProtocolJson.encodeToString(ServerMessageSerializer, decodedReaction),
        )
        assertFalse("futureField" in allJsonKeys(reencodedReaction))
    }

    // --- Malformed frames (§5, §11 drop-and-log posture) ---

    @Test
    fun rejectsAReactMissingEmojiAsMalformed_PROTOCOL5() {
        val frame = """{"type":"react","cell":3}"""
        assertThrows<SerializationException> {
            ProtocolJson.decodeFromString(ClientMessageSerializer, frame)
        }
    }

    @Test
    fun rejectsAReactionWithAMistypedCellAsMalformed_PROTOCOL6() {
        // A structurally wrong `cell` is malformed (§6). The iOS twin uses a quoted number ("3");
        // kotlinx's tree decoder leniently coerces a quoted integer, so the Kotlin twin pins the
        // same rule with a shape no decoder coerces: an object where an integer is required.
        val frame = """{"type":"reaction","userId":"u2","emoji":"🎉","cell":{}}"""
        assertThrows<SerializationException> {
            ProtocolJson.decodeFromString(ServerMessageSerializer, frame)
        }
    }

    // --- The 32-byte shape rule (§9) ---

    @Test
    fun rejectsAnEmptyEmojiAsMalformed_PROTOCOL9() {
        val frame = """{"type":"react","emoji":"","cell":3}"""
        assertThrows<SerializationException> {
            ProtocolJson.decodeFromString(ClientMessageSerializer, frame)
        }
    }

    @Test
    fun rejectsAnEmojiOver32UTF8BytesAsMalformed_PROTOCOL9() {
        // Nine 🎉 graphemes are 36 UTF-8 bytes (4 each), past the 32-byte shape cap.
        val frame = """{"type":"react","emoji":"${"🎉".repeat(9)}","cell":3}"""
        assertThrows<SerializationException> {
            ProtocolJson.decodeFromString(ClientMessageSerializer, frame)
        }
    }

    @Test
    fun acceptsAnEmojiExactlyAtThe32UTF8ByteCap_PROTOCOL9() {
        // Eight 🎉 graphemes are exactly 32 UTF-8 bytes, the inclusive boundary.
        val emoji = "🎉".repeat(8)
        val frame = """{"type":"react","emoji":"$emoji","cell":3}"""
        val decoded = ProtocolJson.decodeFromString(ClientMessageSerializer, frame)
        assertEquals(ClientMessage.React(ReactMessage(emoji = emoji, cell = 3)), decoded)
    }

    @Test
    fun reactionDecoderAppliesTheSameShapeRule_PROTOCOL9() {
        // The inbound notice enforces shape exactly as the outbound command does: one rule, both
        // directions (the codec's single asEmoji, the shared EmojiSerializer).
        val empty = """{"type":"reaction","userId":"u2","emoji":"","cell":3}"""
        assertThrows<SerializationException> {
            ProtocolJson.decodeFromString(ServerMessageSerializer, empty)
        }
        val oversize = """{"type":"reaction","userId":"u2","emoji":"${"🎉".repeat(9)}","cell":3}"""
        assertThrows<SerializationException> {
            ProtocolJson.decodeFromString(ServerMessageSerializer, oversize)
        }
    }
}
